tools/sklearn/dict_h5_pickler.py

```diff
@@ -0,0 +1,510 @@
+    dispatch[complex] = load_primitive
+
+    def load_string(self, data):
+        self.memoize(data)
+        return data
+
+    dispatch[str] = load_string
+
+    def load_unicode(self, data):
+        """
+        `json.load` loads string as unicode in python 2,
+        while some classes don't support unicode, like numpy.dtype
+        """
+        data = str(data)
+        self.memoize(data)
+        return data
+
+    if six.PY2:
+        dispatch[unicode] = load_unicode
+
+    def load_bytes(self, data):
+        data = data.encode('utf-8')
+        self.memoize(data)
+        return data
+
+    def load_list(self, data):
+        return [self.load_all(e) for e in data]
+
+    dispatch[list] = load_list
+
+    def load_tuple(self, data):
+        obj = self.load_all(data)
+        return tuple(obj)
+
+    def load_set(self, data):
+        obj = self.load_all(data)
+        return set(obj)
+
+    def load_dict(self, data):
+        newdict = {}
+        _keys = data[_KEYS]
+        for k in _keys:
+            try:
+                v = data[k]
+            # JSON dumps non-string key to string
+            except KeyError:
+                v = data[str(k)]
+            newdict[k] = self.load_all(v)
+        return newdict
+
+    def load_global(self, data):
+        module = data[0]
+        name = data[1]
+        func = SafePickler.find_class(module, name)
+        self.memoize(func)
+        return func
+
+    def load_reduce(self, data):
+        """
+        Build object
+        """
+        _func = data[_FUNC]
+        func = self.load_all(_func)
+        assert callable(func), "%r" % func
+
+        _args = data[_ARGS][_TUPLE]
+        args = tuple(self.load_all(_args))
+
+        try:
+            obj = args[0].__new__(args[0], * args)
+        except:
+            obj = func(*args)
+
+        _state = data.get(_STATE)
+        if _state:
+            state = self.load_all(_state)
+            setstate = getattr(obj, "__setstate__", None)
+            if setstate:
+                setstate(state)
+            else:
+                assert (type(state) is dict)
+                for k, v in state.items():
+                    setattr(obj, k, v)
+
+        self.memoize(obj)
+        return obj
+
+    def load_np_ndarray(self, data):
+        _dtype = self.load_all(data[_DTYPE])
+        _values = self.load_all(data[_VALUES])
+        obj = numpy.array(_values, dtype=_dtype)
+        return obj
+
+    def load_np_datatype(self, data):
+        _datatype = self.load_all(data[_DATATYPE])
+        _value = self.load_all(data[_VALUE])
+        obj = _datatype(_value)
+        return obj
+
+
+class H5Model:
+    """
+    Convert the dictionary model to h5 and reverse
+    """
+
+    def _create_dataset(self, file_obj, key, value):
+        """
+        Create dataset
+        """
+        try:
+            file_obj.create_dataset(key, data=json.dumps(value))
+        except:
+            file_obj.create_dataset(key, data=value)
```

Comment:
Could you please give some details, about when to use `data=json.dumps(value)`  and when to use `data=value` ?  Thanks.

Reply:
This is particularly useful when an item is `None`. By dumping to json, it becomes `"null"` and while loading `"null"` to `None`